packages/astro/src/integrations/hooks.ts

```diff
@@ -100,6 +100,18 @@ export function getToolbarServerCommunicationHelpers(server: ViteDevServer) {
 	};
 }
 
+// Will match any invalid characters (will be converted to _). We only allow a-zA-Z0-9.-_
+const SAFE_CHARS_RE = /[^\w.-]/g;
+
+export function normalizeInjectedTypeFilename(filename: string, integrationName: string): string {
+	if (!filename.endsWith('.d.ts')) {
+		throw new Error(
+			`Integration ${bold(integrationName)} is injecting a type that does not end with "${bold('.d.ts')}"`,
+		);
+	}
```

Comment:
Is there a world where this API should handle all kinds of filegen? I'm just thinking for instance of the new content intellisense stuff, none of the files being generated are d.ts, right.

Reply:
I had been thinking about an `injectAsset()` API but that will be completely distinct. I think it's fine to keep `injectTypes` focused on dts/references/types and have something else later for the rest (eg. astro db)